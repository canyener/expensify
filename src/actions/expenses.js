import database from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0 
    } = expenseData

    const expense = { description, note, amount, createdAt }

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
}) 

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`/users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value').then(snapshot => {
      const expenses = []

      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })

      dispatch(setExpenses(expenses))
    })
  }
}