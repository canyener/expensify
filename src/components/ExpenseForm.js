import React from 'react'
import moment from 'moment'

import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ?  (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }
  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState( () => ({ description }))
  }
  onNoteChange = (e) => {
    const note = e.target.value
    
    this.setState(() => ({ note }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()

    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }
  render() {
    return (
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input 
            type="text"
            placeholder="Description"
            autoFocus
            className="text-input"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input 
            type="text"
            placeholder="Amount"
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false }
            displayFormat={() => 'DD.MM.YYYY'}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            className="textarea"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <div>
            <button className="button">Save Expense</button>
          </div>
        </form>
    )
  }
}
