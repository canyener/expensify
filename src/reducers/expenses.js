//Expenses Reducer

const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ]
    case 'REMOVE_EXPENSE':
      return state.filter( ({ id }) => id !== action.id)
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if(expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return expense
        }
      })
    case 'SET_EXPENSES':      
      return action.expenses
    default:
      return state
  }
}