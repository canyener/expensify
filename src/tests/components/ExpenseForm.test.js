import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'

import ExpenseForm  from '../../components/ExpenseForm'
import expenses from '../fixtures/expenses'

describe('ExpenseForm', () => {
  describe('#Snapshots', () => {
    test('Should render ExpenseForm correctly with NO expense data', () => {
      const wrapper= shallow(<ExpenseForm />)
      expect(wrapper).toMatchSnapshot()
    })
    
    test('Should render ExpenseForm correctly with given expense data', () => {
      const wrapper = shallow(<ExpenseForm expense={expenses[0]} />)
      expect(wrapper).toMatchSnapshot()
    })
    
    test('Should render error message for invalid form submission', () => {
      const wrapper = shallow(<ExpenseForm />)
    
      expect(wrapper).toMatchSnapshot()
      
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
      })
      
      //expect(wrapper.state('error').length).toBeGreaterThan(0)
      expect(wrapper.state('error')).toBe('Please provide description and amount.')
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('#Functionality', () => {
    test('Should set description on input change', () => {
      const value = 'New description'
      const wrapper = shallow(<ExpenseForm />)
      
      wrapper.find('input').at(0).simulate('change', {
        target: { value }
      })
    
      expect(wrapper.state('description')).toBe(value)
    })
    
    test('Should set note on textarea change', () => {
      const expectedValue = 'New note'
      const wrapper = shallow(<ExpenseForm />)
    
      wrapper.find('textarea').at(0).simulate('change', {
        target: { value: expectedValue }
      })
    
      expect(wrapper.state('note')).toBe(expectedValue)
    })
    
    test('Should set amount if input value is valid', () => {
      const expectedValue = '23.55'
      const wrapper = shallow(<ExpenseForm />)
    
      wrapper.find('input').at(1).simulate('change', {
        target: { value: expectedValue }
      })
    
      expect(wrapper.state('amount')).toBe(expectedValue)
    })
    
    test('Should NOT set amount if input value is invalid', () => {
      const wrapper = shallow(<ExpenseForm />)
    
      wrapper.find('input').at(1).simulate('change', {
        target: { value: '123.456' }
      })
    
      expect(wrapper.state('amount')).toBe('')
    })
    
    test('Should call onSubmit prop for valid form submission', () => {
      const onSubmitSpy = jest.fn()
      const wrapper = shallow(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />)
    
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
      })
    
      expect(wrapper.state('error')).toBe('')
      expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description: expenses[0].description,
        amount: expenses[0].amount,
        note: expenses[0].note,
        createdAt: expenses[0].createdAt
      })
    })
    
    test('Should set new date on date change', () => {
      const now = moment()
      const wrapper = shallow(<ExpenseForm />)
    
      wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(now)
      expect(wrapper.state('createdAt')).toEqual(now)
    })
    
    test('Should set calendarFocused on change', () => {
      const focused = true
      const wrapper = shallow(<ExpenseForm />)
    
      wrapper.find('withStyles(SingleDatePicker)').prop('onFocusChange')({ focused })
      expect(wrapper.state('calendarFocused')).toBe(focused)
    })
  })
})



