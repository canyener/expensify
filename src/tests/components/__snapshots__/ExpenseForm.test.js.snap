// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseForm #Snapshots Should render ExpenseForm correctly with NO expense data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    horizontalMonthPadding={13}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`ExpenseForm #Snapshots Should render ExpenseForm correctly with given expense data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value="Gum"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value="1.95"
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    horizontalMonthPadding={13}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value="Test note for gum"
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`ExpenseForm #Snapshots Should render error message for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    horizontalMonthPadding={13}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`ExpenseForm #Snapshots Should render error message for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide description and amount.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    horizontalMonthPadding={13}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderNavNextButton={null}
    renderNavPrevButton={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;
